linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,330 @@
+
+### Reallocating the Task Array
+
+When there are no unused slots in the circular task array, the task array size is doubled by calling `reallocate_threadpool_array`. New task array memory layout should be in same order of the original array with emilinated gaps between items.  `compress_count` is used to record the number of gaps between `insert_idx` and `consume_idx`. New task array should initialize `consume_idx` as -1 and `insert_idx` as (original array size - gap size - 1) to get the correct index when incremented. If the doubled array size get overflowed for a 32 bits integer, resize failed and return. In this case, the maximum size of task array is 2^32 and the maximum value of `insert_idx` and `consume_idx` is 2^64 which will never get overflowed.
+
+```mermaid
+graph TD
+    A(insert_idx == consume_idx) --> |worker thread runs first| B(consume_idx increment)
+    A--> |main thread runs first| D(no gaps) -->F
+    B--> C(gaps occured)-->E(memmove)-->F(resize done)
+    
+```
+
+## Exposed API
+```C
+
+typedef struct THREADPOOL_TAG* THREADPOOL_HANDLE;
+typedef struct TIMER_INSTANCE_TAG* TIMER_INSTANCE_HANDLE;
+
+#define THREADPOOL_OPEN_RESULT_VALUES \
+    THREADPOOL_OPEN_OK, \
+    THREADPOOL_OPEN_ERROR
+
+MU_DEFINE_ENUM(THREADPOOL_OPEN_RESULT, THREADPOOL_OPEN_RESULT_VALUES)
+
+typedef void (*ON_THREADPOOL_OPEN_COMPLETE)(void* context, THREADPOOL_OPEN_RESULT open_result);
+typedef void (*THREADPOOL_WORK_FUNCTION)(void* context);
+
+MOCKABLE_FUNCTION(, THREADPOOL_HANDLE, threadpool_create, EXECUTION_ENGINE_HANDLE, execution_engine);
+MOCKABLE_FUNCTION(, void, threadpool_destroy, THREADPOOL_HANDLE, threadpool);
+
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+
+MOCKABLE_FUNCTION(, int, threadpool_timer_start, THREADPOOL_HANDLE, threadpool, uint32_t, start_delay_ms, uint32_t, timer_period_ms, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context, TIMER_INSTANCE_HANDLE*, timer_handle);
+
+MOCKABLE_FUNCTION(, int, threadpool_timer_restart, TIMER_INSTANCE_HANDLE, timer, uint32_t, start_delay_ms, uint32_t, timer_period_ms);
+
+MOCKABLE_FUNCTION(, void, threadpool_timer_cancel, TIMER_INSTANCE_HANDLE, timer);
+
+MOCKABLE_FUNCTION(, void, threadpool_timer_destroy, TIMER_INSTANCE_HANDLE, timer);
+
+```
+###  threadpool_create
+
+```C
+MOCKABLE_FUNCTION(, THREADPOOL_HANDLE, threadpool_create, EXECUTION_ENGINE_HANDLE, execution_engine);
+```
+
+`threadpool_create` creates a new threadpool.
+
+**SRS_THREADPOOL_LINUX_11_002: [** If `execution_engine` is `NULL`, `threadpool_create` shall fail and return `NULL`.  **]**
```

Comment:
It looks like the requirements in `interface/devdoc/threadpool_requirements.md ` are incorrectly formatted, probably so a certain other developer could more quickly submit changes.  Could you change the occurrences of "NON_THREADPOOL_" to "SRS_THREADPOOL_" in that file and then fix any build failures that result?